grafast/grafast/src/steps/__inputObject.ts

```diff
@@ -173,6 +173,31 @@ export class __InputObjectStep<
     }
     return !this.inputFields[attrName].step.evalIs(undefined);
   }
+
+  evalKeys(): Array<
+    (keyof TInputType & string) | `${keyof TInputType & number}`
+  > {
+    if (this.inputValues?.kind !== "ObjectValue") {
+      throw new Error("evalKeys must only be called for object types");
+    }
+
+    const keys = new Array<
+      (keyof TInputType & string) | `${keyof TInputType & number}`
+    >();
+    const inputFieldKeys = Object.keys(this.inputFields);
+    for (let i = 0; i < inputFieldKeys.length; i++) {
+      const inputFieldPlan = this.inputFields[inputFieldKeys[i]].step;
+      if (inputFieldPlan.eval() !== undefined) {
```

Comment:
`.eval()` will split the plan for every individual value. Use `evalIs(undefined)` instead, then it will at most have two branches: one for `undefined`, and one for everything else.

```suggestion
      if (!inputFieldPlan.evalIs(undefined)) {
```

Reply:
Note to self: this is required. With `__inputObject` we know that it's an explicit input object (not variable) in the GraphQL document, but the values of each key may still be undefined if they're a variable that isn't supplied and has no default. In these cases we do not wish to return these keys (since the input object is not seen as having those keys set), but that will differ on an operation-to-operation basis, and thus we must evaluate whether or not they are undefined. Note that this implicitly adds constraints for these values; we do not need to explicitly add any constraint for the object itself because the document itself guarantees it will always be present.

Note to self: potential optimization: we should not need to `.evalIs(undefined)` for any input field that is declared as non-nullable, I think?